apps/api_web/lib/api_web/legacy_stops.ex

```diff
@@ -0,0 +1,263 @@
+      "South Station" => ["NEC-2287"],
+      "South Station-01" => ["NEC-2287-01"],
+      "South Station-02" => ["NEC-2287-02"],
+      "South Station-03" => ["NEC-2287-03"],
+      "South Station-04" => ["NEC-2287-04"],
+      "South Station-05" => ["NEC-2287-05"],
+      "South Station-06" => ["NEC-2287-06"],
+      "South Station-07" => ["NEC-2287-07"],
+      "South Station-08" => ["NEC-2287-08"],
+      "South Station-09" => ["NEC-2287-09"],
+      "South Station-10" => ["NEC-2287-10"],
+      "South Station-11" => ["NEC-2287-11"],
+      "South Station-12" => ["NEC-2287-12"],
+      "South Station-13" => ["NEC-2287-13"],
+      "South Weymouth" => ["PB-0158-S"],
+      "Southborough" => ["WML-0274", "WML-0274-01", "WML-0274-02"],
+      "Stoughton" => ["SB-0189-S"],
+      "Swampscott" => ["ER-0128", "ER-0128-01", "ER-0128-02"],
+      "Talbot Avenue" => ["DB-2240", "DB-2240-01", "DB-2240-02"],
+      "TF Green Airport" => ["NEC-1768-03"],
+      "Uphams Corner" => ["DB-2258", "DB-2258-01", "DB-2258-02"],
+      "Wachusett" => ["FR-3338-CS"],
+      "Wakefield" => ["WR-0099", "WR-0099-01", "WR-0099-02"],
+      "Walpole" => ["FB-0191-S"],
+      "Waltham" => ["FR-0098", "FR-0098-01", "FR-0098-S"],
+      "Wareham Village" => ["CM-0493-S"],
+      "Waverley" => ["FR-0074", "FR-0074-01", "FR-0074-02"],
+      "Wedgemere" => ["NHRML-0073", "NHRML-0073-01", "NHRML-0073-02"],
+      "Wellesley Farms" => ["WML-0125", "WML-0125-01", "WML-0125-02"],
+      "Wellesley Hills" => ["WML-0135", "WML-0135-01", "WML-0135-02"],
+      "Wellesley Square" => ["WML-0147", "WML-0147-01", "WML-0147-02"],
+      "West Concord" => ["FR-0219", "FR-0219-01", "FR-0219-02"],
+      "West Gloucester" => ["GB-0296", "GB-0296-01", "GB-0296-02"],
+      "West Hingham" => ["GRB-0162-S"],
+      "West Medford" => ["NHRML-0055", "NHRML-0055-01", "NHRML-0055-02"],
+      "West Natick" => ["WML-0199", "WML-0199-01", "WML-0199-02"],
+      "West Newton" => ["WML-0091-02"],
+      "West Roxbury" => ["NB-0080-S"],
+      "Westborough" => ["WML-0340", "WML-0340-01", "WML-0340-02"],
+      "Weymouth Landing/ East Braintree" => ["GRB-0118-S"],
+      "Whitman" => ["PB-0212-S"],
+      "Wickford Junction" => ["NEC-1659-03"],
+      "Wilmington" => ["NHRML-0152", "NHRML-0152-01", "NHRML-0152-02"],
+      "Winchester Center" => ["NHRML-0078", "NHRML-0078-01", "NHRML-0078-02"],
+      "Windsor Gardens" => ["FB-0166-S"],
+      "Worcester" => ["WML-0442-CS"],
+      "Wyoming Hill" => ["WR-0062", "WR-0062-01", "WR-0062-02"],
+      "Yawkey" => ["WML-0025", "WML-0025-05", "WML-0025-07"]
+    }
+  }
+
+  @doc """
+  Transforms a list of stop IDs as per the given API version and mappings. See the moduledoc for
+  the structure and meaning of mappings.
+  """
+  @spec expand([Stop.id()], String.t(), %{
+          optional(String.t()) => %{optional(Stop.id()) => [Stop.id()]}
+        }) :: [Stop.id()]
+  def expand(stop_ids, api_version, all_mappings \\ @mappings) do
```

Comment:
This works, but one thing to keep in mind is that about a 1/3 of all API requests are going to be hitting this code path (~100/sec), so it's probably worth spending a bit of time thinking about optimization.

One thought right off the bat I had was whether `@mappings` needs to be a map, since the keys aren't guaranteed to be sorted (they are for small maps, but I forget exactly how it works in Erlang and what causes the switch), which is why you're doing the `Enum.sort_by` here. If it was instead a list already, we could skip the `sort` step.

More generally, it's possible to move some of this work to compile time. In particular, we actually know at compile time for a given API version and stop_id what the replacement stops should be. So with a bit of macro magic and computation we could generate a bunch of function heads like:

```ex
def backwards_compatible_stop("South Station", "2020-01-01"), do: ["South Station-01"]
```

Can you maybe poke at that a little bit and see what the implementation complexity is, to see if it's worth it?

Reply:
Good point about the request volume. I'll try benchmarking the current approach vs. `master` to see if there's any concerning differences.

Reply:
From benchmarking it appears this branch adds 1.5µs per request in the best case and 30µs in the worst case (lots of mappings being applied). Our average request time on production is about 3ms, so this branch would increase that by 1% at worst.

I think this is probably acceptable, though I did do a bit of work to see if we could essentially compile the mappings up-front as you suggested. This only gives a 10µs penalty vs. master in the worst case, but also complicates testing of `LegacyStops` since you can't easily dependency-inject the mappings. Overall I'm not sure it's worth it for a gain of 20µs on a 3ms request.